[Feature] add manual numa binding

jasonlizhengjian opened the pull request:
## Purpose

Based off the SGLang implementation using `numactl bind` https://github.com/sgl-project/sglang/pull/14156

## Test Plan

Benchmarks on 1 node of 4xGB200


## Test Result

The most noticeable perf advantage is shown when doing offloading due to the importance of the H2D BW.

commands:
```
serve cmd:
CUDA_VISIBLE_DEVICES=1,2 vllm serve Qwen/Qwen3-4B-Instruct-2507     -dp 2     --enforce-eager     --load-format dummy  --cpu-offload-gb 4 --max-model-len 100000 --numa-node 0 1

bench:
vllm bench serve     --random-input-len 32     --random-output-len 32     --num-prompts 2048     --max-concurrency 256     --request-rate inf     --ignore-eos     --num-warmups 50

```

Here the numa topology optimal setting is `0 1` , which shows a ~1.5x speedup vs all other possible settings.

<details>

```
numa 0 1 
============ Serving Benchmark Result ============
Successful requests:                     2048      
Failed requests:                         0         
Maximum request concurrency:             256       
Benchmark duration (s):                  16.46     
Total input tokens:                      65536     
Total generated tokens:                  65536     
Request throughput (req/s):              124.43    
Output token throughput (tok/s):         3981.70   
Peak output token throughput (tok/s):    5322.00   
Peak concurrent requests:                430.00    
Total token throughput (tok/s):          7963.40   
---------------Time to First Token----------------
Mean TTFT (ms):                          282.46    
Median TTFT (ms):                        239.20    
P99 TTFT (ms):                           674.44    
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          55.46     
Median TPOT (ms):                        55.48     
P99 TPOT (ms):                           68.82     
---------------Inter-token Latency----------------
Mean ITL (ms):                           55.46     
Median ITL (ms):                         43.14     
P99 ITL (ms):                            121.37    
==================================================

numa 1 0
============ Serving Benchmark Result ============
Successful requests:                     2048      
Failed requests:                         0         
Maximum request concurrency:             256       
Benchmark duration (s):                  24.94     
Total input tokens:                      65536     
Total generated tokens:                  65536     
Request throughput (req/s):              82.11     
Output token throughput (tok/s):         2627.40   
Peak output token throughput (tok/s):    3840.00   
Peak concurrent requests:                506.00    
Total token throughput (tok/s):          5254.80   
---------------Time to First Token----------------
Mean TTFT (ms):                          588.39    
Median TTFT (ms):                        565.49    
P99 TTFT (ms):                           960.93    
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          79.79     
Median TPOT (ms):                        78.32     
P99 TPOT (ms):                           108.88    
---------------Inter-token Latency----------------
Mean ITL (ms):                           79.79     
Median ITL (ms):                         66.98     
P99 ITL (ms):                            284.77    
==================================================

numa 0 0 
============ Serving Benchmark Result ============
Successful requests:                     2048      
Failed requests:                         0         
Maximum request concurrency:             256       
Benchmark duration (s):                  24.76     
Total input tokens:                      65536     
Total generated tokens:                  65536     
Request throughput (req/s):              82.72     
Output token throughput (tok/s):         2646.95   
Peak output token throughput (tok/s):    4351.00   
Peak concurrent requests:                381.00    
Total token throughput (tok/s):          5293.91   
---------------Time to First Token----------------
Mean TTFT (ms):                          346.32    
Median TTFT (ms):                        303.21    
P99 TTFT (ms):                           1011.30   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          83.61     
Median TPOT (ms):                        67.51     
P99 TPOT (ms):                           129.71    
---------------Inter-token Latency----------------
Mean ITL (ms):                           83.61     
Median ITL (ms):                         72.37     
P99 ITL (ms):                            196.24    
==================================================


numa 1 1
============ Serving Benchmark Result ============
Successful requests:                     2048      
Failed requests:                         0         
Maximum request concurrency:             256       
Benchmark duration (s):                  24.70     
Total input tokens:                      65536     
Total generated tokens:                  65536     
Request throughput (req/s):              82.90     
Output token throughput (tok/s):         2652.84   
Peak output token throughput (tok/s):    4472.00   
Peak concurrent requests:                379.00    
Total token throughput (tok/s):          5305.69   
---------------Time to First Token----------------
Mean TTFT (ms):                          345.86    
Median TTFT (ms):                        315.48    
P99 TTFT (ms):                           948.39    
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          83.30     
Median TPOT (ms):                        68.30     
P99 TPOT (ms):                           126.27    
---------------Inter-token Latency----------------
Mean ITL (ms):                           83.30     
Median ITL (ms):                         72.35     
P99 ITL (ms):                            219.19    
==================================================

no numa
============ Serving Benchmark Result ============
Successful requests:                     2048      
Failed requests:                         0         
Maximum request concurrency:             256       
Benchmark duration (s):                  24.36     
Total input tokens:                      65536     
Total generated tokens:                  65536     
Request throughput (req/s):              84.09     
Output token throughput (tok/s):         2690.85   
Peak output token throughput (tok/s):    4352.00   
Peak concurrent requests:                448.00    
Total token throughput (tok/s):          5381.70   
---------------Time to First Token----------------
Mean TTFT (ms):                          360.96    
Median TTFT (ms):                        330.20    
P99 TTFT (ms):                           887.80    
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          81.58     
Median TPOT (ms):                        63.79     
P99 TPOT (ms):                           124.58    
---------------Inter-token Latency----------------
Mean ITL (ms):                           81.58     
Median ITL (ms):                         71.90     
P99 ITL (ms):                            209.38    
==================================================
```

</details>

Other Perf Results:

Interestingly, in general, setting to the optimal mask does not have an advantage compared to binding all to the same node. The exact reason why is unknown. 
<details>


```
vllm bench latency \
    --model nvidia/DeepSeek-R1-0528-FP4 \
    --output-len 16 \
    --input-len 8192 \
    --batch-size 4 \
    --tensor-parallel-size 4 \
    --load-format dummy \
    --num_iters_warmup 2 \
    --num_iters 10 \
    --gpu-memory-utilization=0.9 \
    --no-enable-prefix-caching \
    --max-model-len 12000 \
    --max-num-batched-tokens 8192 \
    --enforce-eager
```

```
no numa
Avg latency: 4.731591180500982 seconds
10% percentile latency: 4.493999781300954 seconds
25% percentile latency: 4.502638454505359 seconds
50% percentile latency: 4.506044881505659 seconds
75% percentile latency: 4.9861448467527225 seconds
90% percentile latency: 5.252938730387541 seconds
99% percentile latency: 5.371099184631748 seconds

numa 0 0 1 1 
Avg latency: 4.766906395697151 seconds
10% percentile latency: 4.637488780492276 seconds
25% percentile latency: 4.64759713575404 seconds
50% percentile latency: 4.658970605996728 seconds
75% percentile latency: 4.894723151002836 seconds
90% percentile latency: 5.066911948098277 seconds
99% percentile latency: 5.083919300202542 seconds

numa 0 0 0 0
Avg latency: 4.6389978413004425 seconds
10% percentile latency: 4.4729275049016 seconds
25% percentile latency: 4.47965581324388 seconds
50% percentile latency: 4.558088288504223 seconds
75% percentile latency: 4.821027205998689 seconds
90% percentile latency: 4.9419844798001575 seconds
99% percentile latency: 4.955232554689137 seconds

numa 1 1 1 1
Avg latency: 4.707535313199332 seconds
10% percentile latency: 4.577798735407123 seconds
25% percentile latency: 4.580391513991344 seconds
50% percentile latency: 4.593481449999672 seconds
75% percentile latency: 4.865283180752158 seconds
90% percentile latency: 4.981514163900283 seconds
99% percentile latency: 5.043054415285878 seconds

numa 1 1 0 0
Avg latency: 4.81860344610177 seconds
10% percentile latency: 4.693167993008683 seconds
25% percentile latency: 4.70351736950397 seconds
50% percentile latency: 4.713639230001718 seconds
75% percentile latency: 4.94259596500342 seconds
90% percentile latency: 5.0867359624928215 seconds
99% percentile latency: 5.1469712844536115 seconds

```

`numa 0 0 1 1` is the optimal config according to 

```
$ nvidia-smi topo -m
        GPU0    GPU1    GPU2    GPU3    NIC0    NIC1    NIC2    NIC3    NIC4    NIC5    CPU Affinity    NUMA Affinity   GPU NUMA ID
GPU0     X      NV18    NV18    NV18    NODE    NODE    NODE    SYS     SYS     SYS     0-71    0               N/A
GPU1    NV18     X      NV18    NV18    NODE    NODE    NODE    SYS     SYS     SYS     0-71    0               N/A
GPU2    NV18    NV18     X      NV18    SYS     SYS     SYS     NODE    NODE    NODE    72-143  1               N/A
GPU3    NV18    NV18    NV18     X      SYS     SYS     SYS     NODE    NODE    NODE    72-143  1               N/A
NIC0    NODE    NODE    SYS     SYS      X      NODE    NODE    SYS     SYS     SYS
NIC1    NODE    NODE    SYS     SYS     NODE     X      NODE    SYS     SYS     SYS
NIC2    NODE    NODE    SYS     SYS     NODE    NODE     X      SYS     SYS     SYS
NIC3    SYS     SYS     NODE    NODE    SYS     SYS     SYS      X      NODE    NODE
NIC4    SYS     SYS     NODE    NODE    SYS     SYS     SYS     NODE     X      NODE
NIC5    SYS     SYS     NODE    NODE    SYS     SYS     SYS     NODE    NODE     X
```

but `numa 0 0 0 0` is the best performing in the benchmarks. 

</details>

To effectively use this feature, one should benchmark a variety of numa combinations and empirically find the best one. 


### accuracy

<details>

```
$ vllm serve nvidia/DeepSeek-R1-0528-FP4-V2  --tensor-parallel-size 4 --gpu-memory-utilization=0.9 --numa-node 0 0 1 1

$ python tests/evals/gsm8k/gsm8k_eval.py --port 8000
Running GSM8K evaluation: 1319 questions, 5-shot
Evaluating: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████| 1319/1319 [01:20<00:00, 16.30it/s]

Results:
Accuracy: 0.951
Invalid responses: 0.000
Total latency: 80.950 s
Questions per second: 16.294
Total output tokens: 131743
Output tokens per second: 1627.458


$ vllm serve nvidia/DeepSeek-R1-0528-FP4-V2  --tensor-parallel-size 4 --gpu-memory-utilization=0.9

$ python tests/evals/gsm8k/gsm8k_eval.py --port 8000
Running GSM8K evaluation: 1319 questions, 5-shot
Evaluating: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████| 1319/1319 [01:22<00:00, 16.08it/s]

Results:
Accuracy: 0.951
Invalid responses: 0.000
Total latency: 82.019 s
Questions per second: 16.082
Total output tokens: 131743
Output tokens per second: 1606.247
```

</details>
---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

jasonlizhengjian committed:
add manual numa binding

Signed-off-by: jasonlizhengjian <jasonlizhengjian@gmail.com>

vadiklyutiy commented:
@jasonlizhengjian 
Could you provide some performance number?

vadiklyutiy commented on vllm/config/parallel.py:
```diff
@@ -214,8 +214,12 @@ class is dynamically inherited by the worker class. This is used to inject
+    numa_node: list[int] | None = None
+    """NUMA node for each local GPU. Example: [0, 0, 1, 1] for 4 GPUs
+    on a 2-socket system with GPUs 0-1 on node 0 and GPUs 2-3 on node 1.
+    Use 'nvidia-smi topo -m' to determine GPU-to-NUMA topology."""
```
Maybe we should detect and set it automatically? Don't shift it to customer?

tlrmchlsmth reviewed (commented):
Interesting - @jasonlizhengjian do you have any performance numbers showing improvements from this approach?

Why not just launch vllm with
```
numactl --cpunodebind=N --membind=N vllm serve ...
```

jasonlizhengjian commented:
> Interesting - @jasonlizhengjian do you have any performance numbers showing improvements from this approach?
> 
> Why not just launch vllm with
> 
> ```
> numactl --cpunodebind=N --membind=N vllm serve ...
> ```

yes this is a fair critique, I've not yet found any cases with a perf gain but it might be due to the specific characteristics of the cluster I tested on as the same thing was observed with sglang's numa affinity. 

I will do more benchmarks across different hardwares to see if I can find any cases with a clear perf gain.

jasonlizhengjian committed:
add auto mode and fix dp

Adds --numa-node-auto flag for automatic GPU-to-NUMA topology detection
using NVML, and fixes NUMA binding for Data Parallel workers.

Signed-off-by: Jason Li <jasonlizhengjian@gmail.com>

jasonlizhengjian committed:
fix numa binding for dp with tp/pp

Workers in DP rank 1+ were bound to wrong NUMA nodes because
local_rank was used instead of actual GPU index. Added helper
to compute correct index accounting for data parallelism.

Signed-off-by: Jason Li <jasonl@nvidia.com>

Signed-off-by: jasonlizhengjian <jasonlizhengjian@gmail.com>

jasonlizhengjian commented:
@mgoin please give your thoughts on whether we should have this feature in vllm despite weak benchmark evidence

vadiklyutiy commented:
> Here, numa 0 1 is the optimal config suggested by the topology and is also the fastest in the benchmarks by ~1%.

`CUDA_VISIBLE_DEVICES=4,5` both relate to numa node 1. Did you mix it up?

vadiklyutiy commented:
> Here, numa 0 1 is the optimal config suggested by the topology and is also the fastest in the benchmarks by ~1%.

`CUDA_VISIBLE_DEVICES=2,3` both related to numa node 0 ...

jasonlizhengjian committed:
fix NUMA binding GPU index calculation for EngineCore processes

When spawning EngineCore processes, the NUMA binding was calculating
incorrect GPU indices for DP ranks > 0. The code passed local_dp_rank
as local_rank, but _get_gpu_index() read data_parallel_rank_local from
the config (parent's value), resulting in wrong GPU index calculation.

Added dp_local_rank parameter to configure_subprocess() and
_get_gpu_index() to allow explicit local DP rank specification when
the config doesn't have the correct value yet.

Signed-off-by: Jason Li <jasonlizhengjian@gmail.com>

Signed-off-by: jasonlizhengjian <jasonlizhengjian@gmail.com>

jasonlizhengjian committed:
Merge remote-tracking branch 'origin/main' into numa-manual-bind

# Conflicts:
#	vllm/engine/arg_utils.py
#	vllm/v1/executor/multiproc_executor.py

Signed-off-by: jasonlizhengjian <jasonlizhengjian@gmail.com>

wangshangsam commented:
@jasonlizhengjian are you still working on this feature? I'm just wondering whether and how we can move this forward.

louie-tsai reviewed (commented):
it is good idea to bind GPU workers based on NUMA nodes.  one more feedback, it might be good to also bind specific CPU ids within a NUMA node.  couple benefits:
1. keep GPU workers on same CPU ids will keep those CPU cores frequency high since it got high utilization. with binding to CPU, it might context switch to different CPUs and freq might drop.
2. binding to high frequency CPU cores like Intel PCT(priority core turbo). it should give you even higher frequency and boost the GPU performance.

Harry-Chen reviewed (commented):
The idea is essential for peak performance and I'm happy to see it in vllm. If you are not active recently, I can also help to refine the functionality.
